File: games/game_tigerdragon/DragonTigerConfig/Code/Client/DragonTiger_HelpCFG.cpp
#include "stdafx.h"
#include "tinyxml2/tinyxml2.h"
#include "LuaCfgHelper.h"
#include <iostream>
#include <boost/algorithm/string.hpp>
#include "DragonTiger_HelpCFG.h"
std::auto_ptr<DragonTiger_HelpCFG> DragonTiger_HelpCFG::msSingleton(nullptr);

int DragonTiger_HelpCFG::GetCount()
{
	return (int)mMapData.size();
}

const DragonTiger_HelpCFGData* DragonTiger_HelpCFG::GetData(int HelpID)
{
	auto it = mMapData.find(HelpID);
	if (it != mMapData.end())
	{
		return &it->second;
	}
	return NULL;
}

const std::map<int, DragonTiger_HelpCFGData>& DragonTiger_HelpCFG::GetMapData()
{
	return mMapData;
}

void DragonTiger_HelpCFG::Load()
{
	tinyxml2::XMLDocument xmlDoc;
	std::string content = FileUtils::getInstance()->getStringFromFile("Config/DragonTiger_HelpCFG.xml");
	auto result = xmlDoc.Parse(content.c_str(), content.length());
	if (result != tinyxml2::XML_SUCCESS)
	{
		CCLOGERROR("Result:%d", result);
		CCASSERT(false, "result != tinyxml2::XML_SUCCESS");
		return;
	}
	auto root = xmlDoc.RootElement();
	if (root == NULL)
	{
		CCASSERT(false, "root == NULL");
		return;
	}
	auto element = root->FirstChildElement("Data");
	while (element != NULL)
	{
		DragonTiger_HelpCFGData data;
		data.mHelpID = element->IntAttribute("HelpID");
		data.mCardsName = element->Attribute("CardsName");
		data.mCardsInfo = element->Attribute("CardsInfo");
		data.mCardsTypeStr = element->Attribute("CardsTypeStr");
		{
			const char* Pokers = element->Attribute("Pokers");
			std::vector<std::string> vecPokers;
			boost::split(vecPokers, Pokers, boost::is_any_of(","));
			int temp;
			for (unsigned int i = 0; i < vecPokers.size(); i++)
			{
				if (tinyxml2::XMLUtil::ToInt(vecPokers[i].c_str(), &temp))
				{
					data.mPokers.push_back(temp);
				}
			}
		}
		if (mMapData.find(data.mHelpID) != mMapData.end())std::cout <<"data refind:" << data.mHelpID << std::endl;
		CCASSERT(mMapData.find(data.mHelpID) == mMapData.end(), "data.mHelpID is exists");
		mMapData.insert(std::make_pair(data.mHelpID, data));
		element = element->NextSiblingElement();
	}
	CCLOG("DragonTiger_HelpCFG Loaded. Load Data:%u", mMapData.size());
}

void DragonTiger_HelpCFG::LoadLua()
{
	LuaEngine::getInstance()->executeScriptFile("Config/DragonTiger_HelpCFG");
	lua_State* L = LuaEngine::getInstance()->getLuaStack()->getLuaState();
	lua_getglobal(L, "DragonTiger_HelpCFG");
	CCASSERT(lua_istable(L, -1) == 1, "is not table");
	lua_pushstring(L, "datas");
	lua_gettable(L, -2);
	CCASSERT(lua_istable(L, -1) == 1, "is not table");
	lua_pushnil(L);
	while(lua_next(L, 2))
	{
		CCASSERT(lua_istable(L, -1) == 1, "is not table");
		DragonTiger_HelpCFGData data;
		LuaCfgHelper::readInt(L, "HelpID", data.mHelpID);
		LuaCfgHelper::readString(L, "CardsName", data.mCardsName);
		LuaCfgHelper::readString(L, "CardsInfo", data.mCardsInfo);
		LuaCfgHelper::readString(L, "CardsTypeStr", data.mCardsTypeStr);
		LuaCfgHelper::readVectorInt(L, "Pokers", data.mPokers);
		if (mMapData.find(data.mHelpID) != mMapData.end())std::cout <<"data refind:" << data.mHelpID << std::endl;
		CCASSERT(mMapData.find(data.mHelpID) == mMapData.end(), "data.mHelpID is exists");
		mMapData.insert(std::make_pair(data.mHelpID, data));
		lua_pop(L, 1);
	}
	lua_settop(L, 0);
	CCLOG("DragonTiger_HelpCFG Loaded. Load Data:%u", mMapData.size());
}

void DragonTiger_HelpCFG::Reload()
{
	mMapData.clear();
	Load();
}

DragonTiger_HelpCFG* DragonTiger_HelpCFG::GetSingleton()
{
	if (msSingleton.get() == nullptr)
	{
		msSingleton.reset(new DragonTiger_HelpCFG());
	}
	return msSingleton.get();
}

File: games/game_tigerdragon/DragonTigerConfig/Lua/DragonTiger_HelpCFG.lua
DragonTiger_HelpCFG = {}

function DragonTiger_HelpCFG:getData(key)
	if self.datas == nil then
		return nil
	end
	return self.datas[key]
end

function DragonTiger_HelpCFG:init()
	self.datas = {}
	self.datas[1] = {HelpID = 1, CardsName = "没牛", CardsInfo = "3张相加不为10的倍数 倍率1", CardsTypeStr = "@A", Pokers = {37,40,42,2,52}}
	self.datas[2] = {HelpID = 2, CardsName = "牛二", CardsInfo = "3张相加为10的倍数，另外两张相加各位数为2 倍率1", CardsTypeStr = "02", Pokers = {37,49,52,2,49}}
	self.datas[3] = {HelpID = 3, CardsName = "牛牛", CardsInfo = "3张相加为10的倍数，另外两张相加为10的倍数 倍率5", CardsTypeStr = "00", Pokers = {37,40,49,9,52}}
	self.datas[4] = {HelpID = 4, CardsName = "四炸", CardsInfo = "五张牌中有4张完全一样 倍率7", CardsTypeStr = "<=", Pokers = {31,18,44,52,44}}
	self.datas[5] = {HelpID = 5, CardsName = "五花牛", CardsInfo = "五张牌为花牌(J、Q、K) 倍率8", CardsTypeStr = ">B0", Pokers = {37,50,51,13,24}}
	self.datas[6] = {HelpID = 6, CardsName = "五小牛", CardsInfo = "五张牌均小于5，且相加之和小于10 倍率9", CardsTypeStr = ">?0", Pokers = {28,15,40,3,41}}
end

DragonTiger_HelpCFG:init()
